### ISSUE TITLE
Ember core object different in prod from debug (v2.4.4) results in Assertion

### ISSUE THREAD

**Issue by kingpin2k** [emberjs/ember.js#13304]
The core object has a big difference between the debug and prod version when setting unknown properties that don't have a desc (in the second else below, mandatory setter stuff). It's probably some hidden feature-flag implementation that I don't know about. 

It results in an assertion when used on the DS RecordArray.  I'm guessing this is an ember-data issue, but the code wasn't wrapped in a feature flag and worked in debug, but not prod, so I'm not sure where I should be looking for a fix.  And all of this happens before it ever tries to call back to the server, so not data related.  Maybe I'm missing something obvious here, I'd think this would be affecting more people than just me, so I've probably missed some change somewhere.

http://emberjs.jsbin.com/mumedejoyu/edit?html,js,output

    Assertion Failed: You must use Ember.set() to set the `content` property (of 
    <DS.RecordArray:ember1012>) to ``. Error: Assertion Failed: You must use Ember.set() to 
    set the `content` property (of <DS.RecordArray:ember1012>) to ``.

Debug

    if (desc) {
      desc.set(this, keyName, value);
    } else {
      if (typeof this.setUnknownProperty === 'function' && !(keyName in this)) {
        this.setUnknownProperty(keyName, value);
      } else {
        _emberMetalProperties.defineProperty(this, keyName, null, value); // setup mandatory setter
      }
    }

Prod & min

    if (desc) {
      desc.set(this, keyName, value);
    } else {
      if (typeof this.setUnknownProperty === 'function' && !(keyName in this)) {
        this.setUnknownProperty(keyName, value);
      } else {
        this[keyName] = value;
      }
    }

**Comment by rwjblue** [comment]
1. The `ember-template-compiler.js` script file is not generated with a production version (it always contains assertions).
2. `ember-template-compiler.js` includes all of "ember-metal" (which is where things like this mandatory setter assertion come from).  
3. `ember-template-compiler.js` automatically requires all modules from "ember-metal" when it loads (seeding the `seen` modules listing [here](https://github.com/emberjs/ember.js/blob/master/packages/loader/lib/index.js#L60-L64))
4. `ember.prod.js` loads and redefines all of `ember-metal` modules, using the production version
5. However, at this point, the `seen` registry cache already includes the non-prod versions and re-defining a module doesn't reset that seen cache (it can't because some modules already have a local variable with the original values).

This means, that if you include `ember-template-compiler.js` first and `ember.prod.js` second, you will be using `ember-metal` in development mode (which [includes the mandatory-setter assertion](https://github.com/emberjs/ember.js/blob/master/packages/ember-metal/lib/property_set.js#L94)) but using production version of `ember-runtime` (which will [not be using mandatory-setter](https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/system/core_object.js#L160)).

**Comment by rwjblue** [comment]
**tldr;** Load `ember.prod.js` before `ember-template-compiler.js`: http://rwjblue.jsbin.com/rafini/edit?html,js,output

**Comment by kingpin2k** [comment]
Ahhhhh, I thought it might be some weird setting thing that I wasn't aware of, I'd never seen the issue on CLI, but I'm upgrading a global app right now and they are indeed loaded in the opposite order (like the jsbin).  Thanks good sir.

**Comment by rwjblue** [comment]
Ya, it took me quite a while to track down what was happening :stuck_out_tongue_closed_eyes:

**Comment by kingpin2k** [comment]
Well I appreciate you hunting it down, is your email listed on github one you use? I gotta buy you a beer.
